seagoat/sources/chroma.py

```diff
@@ -22,7 +24,14 @@ def initialize(repository: Repository):
         ),
     )
 
-    chroma_collection = chroma_client.get_or_create_collection(name="code_data")
+    # Check that the user defined provider is in the list of ONNX execution providers
+    if provider is not None and provider in onnxruntime.get_all_providers():
+        chroma_collection = chroma_client.get_or_create_collection(
+            name="code_data",
+            embedding_function=ONNXMiniLM_L6_V2(preferred_providers=[provider]),
```

Comment:
Actually, pretty much looks good to me, the only question in my mind is how we should structure the config file. For example a user might want to use a different embedding function/model, and then different execution providers :thinking:  in this case shouldn't we ask the user to specify that they want to use `ONNXMiniLM_L6_V2`?

Reply:
If the plan is to also support custom embedding functions outside of the support list from `Chroma` I feel like an ideal config would look something like
```yaml
server:
....
client:
   embedding_function:
      name: ONNXMiniLM_L6_V2
      device: CPU <-- Would be useful if the user changed to something like SentenceTransformerEmbeddingFunction from: https://github.com/chroma-core/chroma/blob/main/chromadb/utils/embedding_functions.py#L24C16-L24C16
      onnx_provider: "CPUExecutionProvider" <--- Maybe this would be irrelevant and we could automagically detect which provider to use if CPU or GPU is selected in device field and the model is an onnx model
```
What do you think?

Reply:
I guess the embedding function would be under `server`, but that's a small detail

I like that config format, but looking at it I can't help but think that it could lead to the proliferation of a lot of different config files and incompatibilities/unexpected behaviors between them 🤔 that could lead to some problems if there is community adoption of SeaGOAT :thinking:

So I'm thinking that maybe it's best to just have "presets", so just a single config option. Kinda how you did it, except that instead of specifically about the provider, somehow just have preconfigured embedding functions.

We could have a file like 

**chroma_embedding_presets.py**, sth like this: (pseudo-code)

```python
default = None

def ONNXMiniLM_L6_V2_CPU(...etc):
   return ONNXMiniLM_L6_V2(preferred_providers=[provider])
```

Or I guess, to be a bit more future safe, we could use curried functions (pseudo code):

```python
def default():
    return default

def ONNXMiniLM_L6_V2_CPU(**kwargs):
   def embedding_function():
      return ONNXMiniLM_L6_V2(preferred_providers=[provider])

   return embedding_function
```

That way, if it's really necessary in the future we could enable the user to customize the kwards through the config file. I would normally not add such a things if it's not certain that we need it, but in this case I want to make sure to avoid breaking changes in the config file as much as possible

Reply:
So in the end this is how I would imagine the config file:


```yaml
server:
   chroma:
      embeddingPreset: ONNXMiniLM_L6_V2_CPU
```

What do you think? 

btw in this config example I nested it under `chroma` because I guess later in the future we might allow the user to even disable chroma and use a different database, or perhaps connect to a remote chroma instance, or such

Reply:
if you want we can pair code on implementing this

Reply:
```yaml
server:
   chroma:
      embeddingPreset: ONNXMiniLM_L6_V2_CPU
```
I think if the target device is embedded into the function name it could be quite limiting for the user, as we'd need to infer which arch they are using e.g., `Nvidia, AMD, Intel, Apple`

For PyTorch a user with an Nvidia or AMD GPUs have the same function call. `torch.device('cuda')` (which is super confusing but that's beside the point) but if you wanted to target the GPU on Apple then you'd use `torch.device('mps')` so technically _GPU_ could refer to multiple things here.

Now for the ONNX execution provider (EP) use case we'd need to understand the architecture AND which EP to use. Nvidia has two EPs `CudaExecutionProvider` and `TensorrtExecutionProvider` that target their GPUs. Intel and AMD have two distinct EPs as well with their own supported device matrix. So this could get pretty complex for us to infer what is the right EP to select without letting the user tell us. 

Now I guess SeaGOAT could have all of these as preconfigured embedding functions but I feel that too might be quite messy. 

Those are my thoughts, happy to go with whatever you think is best in this case. 
 
I do really like the option to choose their own vector database as I think that is the safest option long term. 

I'd be glad to do some pair programming implementing a solution!

Reply:
would you please use my calendly link (or otherwise suggest a time) for the pair coding, if you are still interested? https://calendly.com/kantord/30-minute-call @ashishdatta